Rx flow

shamoya opened the pull request:
Please note the flow steering is not tested, as the RQ and TIR are not working for now.
Once the FW will support the RQ creation, the TIR and flow steering code should ~ work.

AFAIU we must have promiscuous flow steering, as we don't have the actual VIRTIO MAC that was configured by QEMU. In switchdev that is ok as the Vswitch will learn the MAC and will configure it uising the represntor to this VF vport.

shamoya committed:
mlx5_prm.h: CREATE_TIR PRM bits addded

User can now create/destroy/modify TIR with DevX.

Signed-off-by: Ido Shamay <idos@mellanox.com>

shamoya committed:
net/mlx5_vdpa: Creating TIR via DEVX

Added TIR creation through DEVX API.
The created TIR is a very simple direct TIR, which points
to RQ of Virtqueue index 0, without any additional features.
The TIR object and number are kept in a new steer_info
struct, which will hold additional steering resources
in the future.

Signed-off-by: Ido Shamay <idos@mellanox.com>

shamoya committed:
net/mlx5_vdpa: set promiscuous steering rule

Configure a promiscuous steering rule through the mlx5dv
flow steering API to steer all traffic of the VF into
the mlx5_vdpa TIR.

Signed-off-by: Ido Shamay <idos@mellanox.com>

shahafsh commented on drivers/net/mlx5/mlx5_vdpa.c:
```diff
@@ -223,8 +287,13 @@ static int mlx5_vdpa_release_virtqs(struct vdpa_priv *priv)
 			}
 			priv->virtq[i].rq_obj = NULL;
 			priv->virtq[i].rqn = 0;
+
```
redundant?

shahafsh commented on drivers/net/mlx5/mlx5_vdpa.c:
```diff
@@ -216,20 +218,76 @@ static int create_tir(struct vdpa_priv *priv)
+	struct mlx5dv_flow_matcher_attr dv_attr = {
+		.type = IBV_FLOW_ATTR_NORMAL,
+		.match_mask = (void *)&empty_match,
+
```
redundant blank line.

shahafsh commented on drivers/net/mlx5/mlx5_vdpa.c:
```diff
@@ -216,20 +218,76 @@ static int create_tir(struct vdpa_priv *priv)
+	int ret = 0;
+	struct ibv_flow *promisc_flow = NULL;
+	struct mlx5dv_flow_matcher *matcher = NULL;
+	struct mlx5_flow_dv_match_params empty_match = {
```
better to use "val" instead of match on this context. it is confusing v.s. the matcher.

shahafsh reviewed (approved):
@shamoya very nice. 

see very minor comments

shamoya committed:
net/mlx5_vdpa: Release rx_steer resources before Virtqueues

Since releasing verbs resources should be in LIFO order,
move the the release_rx_steer function (which is allocated last)
to the front of the release function.

Signed-off-by: Ido Shamay <idos@mellanox.com>

shamoya commented:
Thanks @shahafsh 
Fixed comments and pushed
